DTML.EduBot/Dialogs/LessonDialog.cs

```diff
@@ -84,36 +74,158 @@ public LessonDialog(Lesson lesson)
+        {
+            var message = await result;
+            StudentResponse studentResponse = new StudentResponse(message.Text);
+
+            // TODO think about how to deal with the bot response right after typed answer
+            await CheckAnswerAsync(context, studentResponse, "Great, now type anything to continue", this.PronounceLearnedPhrase);
+        }
+
+        private async Task CheckAnswerAsync(IDialogContext context, StudentResponse studentResponse, string botResponse, ResumeAfter<IMessageActivity> resume)
+        {
             // TODO: transform into strongly typed.
-            dynamic response = message.Value;
+            
+
             var topic = lesson.Topics.ElementAtOrDefault(lesson.currentTopic);
             if (topic == null)
             {
                 return;
             }
 
-            StudentResponse studentResponse = StudentResponse.FromDynamic(response);
-
-            if (studentResponse.Answer.Equals(topic.CorrectAnswer, StringComparison.InvariantCultureIgnoreCase))
+            if (studentResponse.Answer != null && studentResponse.Answer.Equals(topic.CorrectAnswer, StringComparison.InvariantCultureIgnoreCase))
             {
-                await context.PostAsync("You have the right answer! Moving on to the next question.");
-                lesson.currentTopic++;
-                await this.StartAsync(context);
+                // TODO fix the bug of showing bot respones after typed solution
+                await context.PostAsync(botResponse);
+                context.Wait(resume);
             }
             else
             {
-                await context.PostAsync("Please try again");
+                await context.PostAsync(topic.WrongAnswerBotResponse);
                 await this.StartAsync(context);
             }
         }
+
+        private async Task PronounceLearnedPhrase(IDialogContext context, IAwaitable<IMessageActivity> result)
+        {
+            var topic = lesson.Topics.ElementAtOrDefault(lesson.currentTopic);
+            if (topic == null)
+            {
+                return;
+            }
+
+            await this.PostAudioInstruction(context, topic);
+
+            await this.WrapUpCurrentTopic(context, topic);
+        }
+
+        private async Task PostAudioInstruction(IDialogContext context, Topic topic)
+        {
+            await context.PostAsync(topic.PronounciationPhrase);
+
+            // TODO: integrate with text to speech API 
+        }
+
+        private Task WrapUpCurrentTopic(IDialogContext context, Topic topic)
+        {
+            PromptDialog.Choice(
+                context,
+                this.AfterWrapUpCurrentTopic,
+                topic.WrapUpPhrases,
+                "This marks the end of current topic!",
+                "I am sorry but I didn't understand that. I need you to select one of the options below",
+                attempts: topic.WrapUpPhrases.Count());
```

Comment:
Because user can type in something else other than "I got it" or "say it again". In that case, I don't think we should allow user to keep retry, and instead the number of wrap up phrases seems like a good attempt max number to me.